rcl/src/rcl/init.c

```diff
@@ -0,0 +1,172 @@
+  context->impl->argc = argc;
+  context->impl->argv = NULL;
+  if (0 != argc && argv != NULL) {
+    context->impl->argv = (char **)allocator.allocate(sizeof(char *) * argc, allocator.state);
+    RCL_CHECK_FOR_NULL_WITH_MSG(
+      context->impl->argv,
+      "failed to allocate memory for argv",
+      fail_ret = RCL_RET_BAD_ALLOC; goto fail);
+    memset(context->impl->argv, 0, sizeof(char *) * argc);
+    int64_t i;
+    for (i = 0; i < argc; ++i) {
+      size_t argv_i_length = strlen(argv[i]);
+      context->impl->argv[i] = (char *)allocator.allocate(argv_i_length, allocator.state);
+      RCL_CHECK_FOR_NULL_WITH_MSG(
+        context->impl->argv[i],
+        "failed to allocate memory for string entry in argv",
+        fail_ret = RCL_RET_BAD_ALLOC; goto fail);
+      memcpy(context->impl->argv[i], argv[i], argv_i_length);
+    }
+  }
+
+  // Parse the ROS specific arguments.
+  ret = rcl_parse_arguments(argc, argv, allocator, &context->global_arguments);
+  if (RCL_RET_OK != ret) {
+    fail_ret = ret;
+    RCUTILS_LOG_ERROR_NAMED(ROS_PACKAGE_NAME, "Failed to parse global arguments");
+    goto fail;
+  }
+  // Update the default log level if specified in arguments.
+  if (context->global_arguments.impl->log_level >= 0) {
+    rcutils_logging_set_default_logger_level(context->global_arguments.impl->log_level);
+  }
+
+  // Set the instance id.
+  uint64_t next_instance_id = rcutils_atomic_fetch_add_uint64_t(&__rcl_next_unique_id, 1);
+  if (0 == next_instance_id) {
+    // Roll over occurred, this is an extremely unlikely occurrence.
+    RCL_SET_ERROR_MSG("unique rcl instance ids exhausted");
+    // Roll back to try to avoid the next call succeeding, but there's a data race here.
+    rcutils_atomic_store(&__rcl_next_unique_id, -1);
+    goto fail;
+  }
+  rcutils_atomic_store((atomic_uint_least64_t *)(&context->instance_id_storage), next_instance_id);
+
+  // Initialize rmw_init.
+  context->impl->rmw_context = rmw_get_zero_initialized_context();
+  rmw_ret_t rmw_ret = rmw_init(
+    &(context->impl->init_options.impl->rmw_init_options),
+    &(context->impl->rmw_context));
+  if (RMW_RET_OK != rmw_ret) {
+    RCL_SET_ERROR_MSG(rmw_get_error_string().str);
+    fail_ret = RCL_RET_ERROR;
+    goto fail;
+  }
+
+  // Store the allocator.
+  context->impl->allocator = allocator;
+
+  return RCL_RET_OK;
+fail:
+  __cleanup_context(context);
+  return fail_ret;
+}
+
+rcl_ret_t
+rcl_shutdown(rcl_context_t * context)
+{
+  RCUTILS_LOG_DEBUG_NAMED(
+    ROS_PACKAGE_NAME,
+    "Shutting down ROS client library, for context at address: %p", context);
+  RCL_CHECK_ARGUMENT_FOR_NULL(context, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    context->impl, "context is zero-initialized", return RCL_RET_INVALID_ARGUMENT);
+  if (!rcl_context_is_valid(context)) {
+    RCL_SET_ERROR_MSG("rcl_shutdown already called on the given context");
+    return RCL_RET_ALREADY_SHUTDOWN;
+  }
+
```

Comment:
also call `rmw_shutdown()`?

Reply:
Definitely, fixed that in latest commits.